### PR TITLE
added first version of didman api

### PR DESCRIPTION
**Pull request by woutslakhorst**

### CODE REVIEW

File: docs/_static/didman/v1.yaml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        example:
+          - "did:nuts:1234#service-x"
+        schema:
+          type: string
+    delete:
+      summary: Remove a service from a DID Document.
+      description: |
+        Remove a service from a DID Document. If it's a concrete service and still referenced by other service, it'll return a 409.
+        If the service can't be found, it'll return a 404.
+      operationId: "deleteEndpoint"
```
Comment (reinkrul) [inline docs/_static/didman/v1.yaml:55]: `deleteService`

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: docs/_static/didman/v1.yaml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  schemas:
+    EndpointCreateRequest:
+      type: object
+      description: A combination of type and URL.
+      required:
+        - type
+        - endpoint
+      properties:
+        type:
+          description: type of the endpoint. May be freely choosen.
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply docs/_static/didman/v1.yaml:71]: bolt specific, no limitations from tech

File: docs/_static/didman/v1.yaml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        required: true
+        example:
+          - "did:nuts:1234#service-x"
+        schema:
+          type: string
+    delete:
+      summary: Remove a service from a DID Document.
+      description: |
+        Remove a service from a DID Document. If it's a concrete service and still referenced by other service, it'll return a 409.
+        If the service can't be found, it'll return a 404.
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply docs/_static/didman/v1.yaml:54]: The choice has been made to use the `default` for all services and add specific returns codes to the description. This is way easier to read when published as swagger-doc. But I'll make it easier to read in the description.

File: didman/interface.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	"net/url"
+
+	ssi "github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/go-did/did"
+)
+
+// Didman groups all high-level methods for manipulating DID Documents
+type Didman interface {
+	// AddEndpoint adds a service to a DID Document. The serviceEndpoint is set to the given URL.
+	AddEndpoint(id did.DID, serviceType string, u url.URL) error
```
Comment (stevenvegt) [inline didman/interface.go:32]: Document the possible errors here as well.

File: didman/api/v1/api.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+		return core.NewProblem(problemTitleDeleteService, http.StatusBadRequest, err.Error())
+	}
+
+	if err = w.Didman.DeleteService(*id); err != nil {
+		logging.Log().WithError(err).Warn(problemTitleDeleteService)
+		if errors.Is(err, types.ErrNotFound) {
+			return core.NewProblem(problemTitleDeleteService, http.StatusNotFound, err.Error())
+		}
+		if errors.Is(err, types.ErrDIDNotManagedByThisNode) {
+			return core.NewProblem(problemTitleAddEndpoint, http.StatusBadRequest, err.Error())
```
Comment (reinkrul) [inline didman/api/v1/api.go:117]: this should be `problemTitleDeleteService`?

File: didman/api/v1/api.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+		}
+		if errors.Is(err, types.ErrDeactivated) {
+			return core.NewProblem(problemTitleAddEndpoint, http.StatusConflict, err.Error())
```
Comment (reinkrul) [inline didman/api/v1/api.go:120]: same here

Comment (reinkrul) [inline didman/api/v1/api.go:120]: same here

File: didman/api/v1/api.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		}
+		if errors.Is(err, didman.ErrServiceInUse) {
+			return core.NewProblem(problemTitleDeleteService, http.StatusConflict, err.Error())
+		}
+		return core.NewProblem(problemTitleAddEndpoint, http.StatusInternalServerError, err.Error())
```
Comment (reinkrul) [inline didman/api/v1/api.go:125]: also

**Review by stevenvegt — APPROVED** [review]

File: didman/didman.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+		vdr:         vdr,
+	}
+}
+
+func (d *didman) Name() string {
+	return ModuleName
+}
+
+func (d *didman) AddEndpoint(id did.DID, serviceType string, u url.URL) error {
+	doc, meta, err := d.docResolver.Resolve(id, nil)
```
Comment (reinkrul) [inline didman/didman.go:66]: logging (debug at start, info at end)

File: didman/didman.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+	// construct service with correct ID
+	service := constructService(id, serviceType, u)
+	doc.Service = append(doc.Service, service)
+
+	return d.vdr.Update(id, meta.Hash, *doc, nil)
+}
+
+func (d *didman) DeleteService(serviceID ssi.URI) error {
+	id, err := did.ParseDID(serviceID.String())
```
Comment (reinkrul) [inline didman/didman.go:86]: also needs logging

File: didman/didman.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	id.Fragment = ""
+
+	doc, meta, err := d.docResolver.Resolve(*id, nil)
+	if err != nil {
+		return err
+	}
+
+	// check for existing use
+	if err = d.store.Iterate(func(doc did.Document, metadata types.DocumentMetadata) error {
```
Comment (reinkrul) [inline didman/didman.go:98]: will this be performant enough?

File: didman/didman.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+	return service
+}
+
+type compoundService map[string]string
+
+func referencesService(doc did.Document, serviceID ssi.URI) bool {
+	id := serviceID.String()
+	for _, s := range doc.Service {
+		cs := compoundService{}
```
Comment (reinkrul) [inline didman/didman.go:142]: copy/paste from `auth/services/util.go:48`?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: didman/didman.go (hunk 11 above)
Comment (woutslakhorst) [reply didman/didman.go:142]: Ahhh the `type`. Maybe I can move it to the VDR as public type.

File: didman/didman.go (hunk 10 above)
Comment (woutslakhorst) [reply didman/didman.go:98]: No, but adding indices on the in-memory implementation would be a waste of time. When the implementation changes to a persistent store, that store hopefully has indices.

**Review by reinkrul — APPROVED** [review]

**Pull request merged**